File: app/static/CustomListener.js
var antlr4 = require('antlr4/index');

// This class defines a complete listener for a parse tree produced by ExprParser.
function Listener() {
	antlr4.tree.ParseTreeListener.call(this);
    this.nodeMap = {};
    this.nodeEdges = [];
    this.trivialRoot = null;
    this.crnt_id = -1;
    this.makeCytoNode = makeCytoNode;
	return this;
}

Listener.prototype = Object.create(antlr4.tree.ParseTreeListener.prototype);
Listener.prototype.constructor = Listener;

Listener.prototype.enterEveryRule = function(ctx) {
    //console.log(`entering node ${ctx.constructor.name}`);
    console.log(ctx);
    this.crnt_id++;
    ctx.cytoId = this.crnt_id;
    this.nodeMap[this.crnt_id] = this.makeCytoNode(ctx);
    // this is redundant..
    if (isTrivial(ctx) && !this.trivialRoot) this.trivialRoot = ctx.parentCtx.cytoId;
    if (!isTrivial(ctx) && this.trivialRoot) {
        this.nodeEdges.push({
            data: {
                source: this.trivialRoot,
                target: ctx.cytoId
            },
            classes: "collapsed"
        });
        this.trivialRoot = null;
    }
};

Listener.prototype.exitEveryRule = function(ctx) {
    var edges = this.nodeEdges;
    ctx.children.forEach(function(v){
        edges.push({
            data: {
                id: `${ctx.cytoId}-${v.cytoId}`, 
                source: ctx.cytoId,
                target: v.cytoId
            }
        });
    });

    //console.log('exiting node');
};

Listener.prototype.visitTerminal = Listener.prototype.enterEveryRule;
Listener.prototype.visitErrorNode = Listener.prototype.enterEveryRule;

function makeCytoNode(ctx){
    var start = ctx.start ? ctx.start : ctx.symbol;
    var stop  = ctx.stop  ? ctx.stop  : ctx.symbol;
    var data = {
        id: this.crnt_id,
        text: ctx.getText(),
        // Note, this is because antlr4 adds Context to end of each
        // name by default. This approach could break if Context is part of
        // nodes actual name
        name: ctx.constructor.name.replace("Context", ""),
        trivial: isTrivial(ctx),
        lineInfo: {
            col_start: start.start,
            line_start: start.line,
            col_end: stop.stop,
            line_end: stop.line
        }
    }

    if (ctx.constructor.name == "TerminalNodeImpl") var classes = "terminal";
    else if (ctx.constructor.name == "ErrorNodeImpl") var classes = "error";
    else var classes = "context";

    return {data, classes}
};

function tokenToCytoNode(ctx){
    var data = {
        id: this.crnt_id,
        text: ctx.getText(),
        name: ctx.constructor.name,
        lineInfo: {
            col_start: ctx.symbol.column,
            line_start: ctx.symbol.line,
            col_end: ctx.symbol.column,
            line_end: ctx.symbol.line
        }
    }
}

function isTrivial(ctx){
    if (   
           ctx.parentCtx != null 
        && ctx.parentCtx.children.length == 1 
        && ctx.children != null
        && ctx.children.length == 1
        && ctx.children[0].constructor.name != "TerminalNodeImpl"
        ) return true
    else return false
}

// visitErrorNode
// visitTerminal
//
exports.Listener = Listener;

File: app/CustomListener.py
from antlr4.tree import Tree
from antlr4 import InputStream, CommonTokenStream


class Listener(Tree.ParseTreeListener):
    """This class defines a complete listener for a parse tree produced by ExprParser."""

    def __init__(self):
        super(Listener, self).__init__()
        self.node_map = {}
        self.node_edges = []
        self.trivial_root = None
        self.current_id = -1

    def enterEveryRule(self, ctx):
        self.current_id += 1
        ctx.cyto_id = self.current_id
        self.node_map[self.current_id] = self.make_cyto_node(ctx)
        # this is redundant..
        if self.is_trivial(ctx) and not self.trivial_root:
            self.trivial_root = ctx.parentCtx.cyto_id
        if not self.is_trivial(ctx) and self.trivial_root:
            self.node_edges.append(
                {
                    "data": {"source": self.trivial_root, "target": ctx.cyto_id},
                    "classes": "collapsed",
                }
            )
            self.trivial_root = None

    def exitEveryRule(self, ctx):
        if not ctx.children:
            return

        edges = self.node_edges
        for child in ctx.children:
            edges.append(
                {
                    "data": {
                        "id": "%s-%s" % (ctx.cyto_id, child.cyto_id),
                        "source": ctx.cyto_id,
                        "target": child.cyto_id,
                    }
                }
            )

    def visitTerminal(self, node):
        self.enterEveryRule(node)

    def visitErrorNode(self, node):
        self.enterEveryRule(node)

    def make_cyto_node(self, ctx):
        name = ctx.__class__.__name__
        start = ctx.start if hasattr(ctx, "start") else ctx.symbol
        stop = ctx.stop if hasattr(ctx, "stop") else ctx.symbol
        data = {
            "id": self.current_id,
            "text": ctx.getText(),
            # Note, this is because antlr adds Context to end of each
            # name by default. This approach could break if Context is part of
            # nodes actual name
            "name": name.replace("Context", ""),
            "trivial": self.is_trivial(ctx),
            "lineInfo": {
                "col_start": start.start,
                "line_start": start.line,
                "col_end": stop.stop,
                "line_end": stop.line,
            },
        }

        if name == "TerminalNodeImpl":
            classes = "terminal"
        elif name == "ErrorNodeImpl":
            classes = "error"
        else:
            classes = "context"

        return {"data": data, "classes": classes}

    @staticmethod
    def is_trivial(ctx):
        if (
            ctx.parentCtx is not None
            and len(ctx.parentCtx.children) == 1
            and hasattr(ctx, "children")
            and len(ctx.children) == 1
            and ctx.children[0].__class__.__name__ != "TerminalNodeImpl"
        ):
            return True
        else:
            return False


def parse_from_grammar(grammar, data, start):
    chars = InputStream(data)
    lexer = grammar.Lexer(chars)
    tokens = CommonTokenStream(lexer)
    parser = grammar.Parser(tokens)
    parser.buildParseTrees = True

    tree = getattr(parser, start)()
    listener = Listener()
    Tree.ParseTreeWalker.DEFAULT.walk(listener, tree)

    elements = {
        "nodes": [listener.node_map[key] for key in listener.node_map.keys()],
        "edges": listener.node_edges,
    }
    return elements


if __name__ == "__main__":
    from app.__init__ import grammars

    raw_ast = parse_from_grammar(
        grammars["plsql"], "SELECT id FROM artists WHERE id > 100", "sql_script"
    )

    print(raw_ast)
